### DISCUSSION TITLE
Set cookies when returning a RedirectResponse

### DISCUSSION THREAD

**Discussion by coltonbh** [fastapi/fastapi#7077] [Questions] [answered]
### Example

```Python
from fastapi import FastAPI
from fastapi.responses import Response, RedirectResponse

app = FastAPI()

@app.get("/")
def root():
    return {"Hello": "World"}

@app.get("/cookieset")
def cookie_set(response: Response):
    response.set_cookie(key="works", value="here is your data")
    return "My cookies got set!"

@app.get("/cookiefail")
def cookie_fail(response: Response):
    response.set_cookie(key="works", value="here is your data")
    return RedirectResponse("/")

@app.get("/cookieset2")
def cookie_set2(response: Response):
    # This works, but appears to fall short of what the documentation claims, noted below.
    response.set_cookie(key="works", value="here is your data")
    response.status_code = 307
    response.headers["location"] = "/"
    return response

@app.get("/cookieset3", response_class=RedirectResponse)
def cookie_set3() -> RedirectResponse:
    # Proposed by @mcauto in comments below; sets cookie correctly. Pretty close to "ideal" behavior, I think the best behavior would be /cookieset4
    response = RedirectResponse(url="/")
    response.set_cookie(key="works", value="here is your data", domain="127.0.0.1")
    return response

@app.get("/cookieset4", response_class=RedirectResponse)
def cookie_set4(response: RedirectResponse) -> RedirectResponse:
    response.set_cookie(key="works", value="here is your data", domain="127.0.0.1")
    return response
```

### Description

* Open the browser and call the endpoint `/cookieset`.
* Note the cookie is properly set in localstorage
* Delete the cookie from local storage.

* Open the browser and call the endpoint `/cookiefail`.
* Note the cookie is NOT set when the response is a redirect response (check local storage for cookies, it is empty)

* The final `/cookieset2` endpoint gives the correct behaviors, but the documentation [here](https://fastapi.tiangolo.com/advanced/response-cookies/#response-cookies) appears to suggest that setting cookies on the response argument will be correctly incorporated into the final response. 

### The solution you would like

I would like cookies to be set when returning a RedirectResponse.

### Environment

* OS:  macOS:
* FastAPI Version 0.62.0:
* Python version: 3.9.0

**Comment by Mause** [comment]
If you look at the documentation directly below where you linked, it shows that if you are returning a response instance, you must set the cookie on that instance:

https://fastapi.tiangolo.com/advanced/response-cookies/#return-a-response-directly

**Comment by mcauto** [comment]
@coltonbh 

Did you test your domain `localhost`?

**Comment by coltonbh** [comment]
@Mause it appears you are correct that the documentation states this behavior! I missed that. I still think this would be a nice feature for people as the behavior is a bit unexpected and I spent some time trying to diagnose it, but it is evidently documented. Should I close given that this behavior is in accordance with the documentation or should I leave it open as a feature request?

@mcauto what exactly do you mean? Yes, I was running this locally on my machine on `localhost`. Not sure what you mean "test your domain?" Can you clarify a bit? Thanks!

**Comment by ycd** [comment]
@coltonbh Sure, leave it as a feature request also you might want to rename the question with the `[FEATURE]` label.

**Comment by mcauto** [comment, marked as answer]
@coltonbh 

```python
@app.get("/cookieset2", response_class=RedirectResponse)
def cookie_set2() -> RedirectResponse:
    response = RedirectResponse(url="/")
    response.set_cookie(key="works", value="here is your data", domain="127.0.0.1")
    return response
```

**Comment by coltonbh** [comment]
@mcauto your example also does not set the cookie. I can add that to my question as another failing case.

**Comment by mcauto** [comment]
@coltonbh  

It's works.

![image](https://user-images.githubusercontent.com/15080806/101437971-aa2c5400-3954-11eb-9000-cd9d066716ad.png)

**Comment by coltonbh** [comment]
@mcauto you are correct! I think I tested the wrong endpoint. Thanks!

**Comment by northtree** [comment]
@coltonbh The solution from @mcauto is good. Could you close this issue. Thanks.

**Comment by coltonbh** [comment]
@northtree happy to! @ycd had asked for this to remain open as a feature request as per our conversation above. Confirm desire to close this?

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

If that solves the original problem, then you can close this issue :heavy_check_mark:

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by coltonbh** [comment]
This could be left open as a feature request, i.e., to have the `RedirectResponse` set cookies; however, my initial confusion was addressed. Will close.

**Comment by isadri** [comment]
Hi @mcauto. Why it does not work in my case.
I have this path operation function that will set a cookie then redirect to another endpoint:
```python
@app_mock.get('/authorize')
async def authorize(
    redirect_uri: Annotated[str, Query()],
    state: Annotated[str, Query()],
    code: Annotated[str, Depends(lambda: OAUTH_CODE)],
    request: Request,
):
    uri: str = f'{redirect_uri}?code={code}&state={state}'
    response = RedirectResponse(uri)
    state_cookie = request.cookies['state']
    response.set_cookie(key='state', value=state_cookie)
    return response
```
The other path operation function associated with the redirect uri has this cookie parameter, here is some snippet:
```python
state_cookie: Annotated[str, Cookie(alias='state')],
```
When I run pytest I got this error:
```json
 {'type': 'missing', 'loc': ('cookie', 'state'), 'msg': 'Field required', 'input': None}
```
